### DISCUSSION TITLE
Getters on context - it feels forbidden

### DISCUSSION THREAD

**Discussion by Asderex** [statelyai/xstate#3142] [Q&A] [answered]
Hi,  Let me preface this by saying I'm aware this might be a little off piste with xstates way of doing things but I was hoping to have additional values on context whos values are derived from other existing context values.  The obvious way for me to do this seemed to be using getters.  For some reason when I pass an object with getters into context the getters are hardcoded when the context is provided.  Is this by design? and if so is there a recommended way to have second order context other than manually deriving it each time from context?

I have a basic codesandbox example available [here](https://codesandbox.io/s/getter-on-xstate-context-dr5612?file=/src/index.js):
Note in the logs the fullName doesn't change despite the first name being updated. This is a very simplified example but it illustrates my point above.

I've noted that instead of getters I can add functions just fine and call them with the added (). I was just curious why the getters appear to break?

On a completely unrelated note - awesome job with Stately.ai : Loving it!!!!

**Comment by Andarist** [comment, marked as answer]
`assign` is always spreading its result on the existing context object and that "invokes" getters:
```js
var context = { a: 1 }
var incoming = { get b() { return 2 } }
var final = { ...context, ...incoming } // this ends up being { a: 1, b: 2 } and NOT { a: 1, get b() { return 2 } }
```

> I've noted that instead of getters I can add functions just fine and call them with the added ()

I would be cautious about this - but I think that if only your functions will use `this` to accent the current value of the context and not the closure then it should work.

>  and if so is there a recommended way to have second order context other than manually deriving it each time from context?

The usual answer for this is to just use functions aka selectors on your context to derive those values. I understand that this might not be the best answer from your PoV but given the design and language constraints, this seems to be the most straightforward way of doing such things.

**Reply by Asderex** [reply]
Thanks Andarist - thats a clear explanation. Appreciate your time and the suggested solution.
